cpp/src/arrow/compute/kernels/select_k_test.cc

```diff
@@ -0,0 +1,714 @@
+
+  std::shared_ptr<Array> Generate(uint64_t count, double null_prob, double nan_prob = 0) {
+    return generator_.Float64(count, std::numeric_limits<CType>::min(),
+                              std::numeric_limits<CType>::max(), null_prob, nan_prob);
+  }
+};
+
+template <>
+class Random<Decimal128Type> : public RandomImpl {
+ public:
+  explicit Random(random::SeedType seed,
+                  std::shared_ptr<DataType> type = decimal128(18, 5))
+      : RandomImpl(seed, std::move(type)) {}
+};
+
+template <typename ArrowType>
+class RandomRange : public RandomImpl {
+  using CType = typename TypeTraits<ArrowType>::CType;
+
+ public:
+  explicit RandomRange(random::SeedType seed)
+      : RandomImpl(seed, TypeTraits<ArrowType>::type_singleton()) {}
+
+  std::shared_ptr<Array> Generate(uint64_t count, int range, double null_prob) {
+    CType min = std::numeric_limits<CType>::min();
+    CType max = min + range;
+    if (sizeof(CType) < 4 && (range + min) > std::numeric_limits<CType>::max()) {
+      max = std::numeric_limits<CType>::max();
+    }
+    return generator_.Numeric<ArrowType>(count, min, max, null_prob);
+  }
+};
+
+TYPED_TEST_SUITE(TestSelectKRandom, SelectKableTypes);
+
+TYPED_TEST(TestSelectKRandom, RandomValues) {
+  Random<TypeParam> rand(0x61549225);
+  int length = 100;
+  for (auto null_probability : {0.0, 0.1, 0.5, 1.0}) {
+    // Try n from 0 to out of bound
+    for (int n = 0; n <= length; ++n) {
+      auto array = rand.Generate(length, null_probability);
+      this->AssertTopKArray(array, n);
+      this->AssertBottomKArray(array, n);
+    }
+  }
+}
+
+template <SortOrder order>
+struct TestSelectKWithChunkedArray : public ::testing::Test {
+  TestSelectKWithChunkedArray()
+      : sizes_({0, 1, 2, 4, 16, 31, 1234}),
+        null_probabilities_({0.0, 0.1, 0.5, 0.9, 1.0}) {}
+
+  void Check(const std::shared_ptr<DataType>& type,
+             const std::vector<std::string>& values, int64_t k,
+             const std::string& expected) {
+    std::shared_ptr<Array> actual;
+    ASSERT_OK(this->DoSelectK(type, values, k, &actual));
+    ValidateOutput(actual);
+
+    ASSERT_ARRAYS_EQUAL(*ArrayFromJSON(type, expected), *actual);
+  }
+
+  void Check(const std::shared_ptr<DataType>& type,
+             const std::shared_ptr<ChunkedArray>& values, int64_t k,
+             const std::string& expected) {
+    ASSERT_OK_AND_ASSIGN(auto actual, SelectK<order>(*values, k));
+    ValidateOutput(actual);
+    ASSERT_ARRAYS_EQUAL(*ArrayFromJSON(type, expected), *actual);
+  }
+
+  Status DoSelectK(const std::shared_ptr<DataType>& type,
+                   const std::vector<std::string>& values, int64_t k,
+                   std::shared_ptr<Array>* out) {
+    ARROW_ASSIGN_OR_RAISE(*out, SelectK<order>(*(ChunkedArrayFromJSON(type, values)), k));
+    PrettyPrint(**out, {}, &std::cerr);
+    return Status::OK();
+  }
+  std::vector<int32_t> sizes_;
+  std::vector<double> null_probabilities_;
+};
+
+struct TestTopKWithChunkedArray
```

Comment:
Why not make these checks above in the typed tests?

Reply:
You can make the 'base' overload of the test helpers take Datum and then use ChunkedArrayFromJSON in the tests themselves, so that you don't have to write too many overloads.

Reply:
Also, you can probably make one typed test suite that covers all the numeric and most of the temporal types in one go, since the temporal types generally are numeric underneath.